Improve fullscreen, orientation and PiP handling

ThreeFive-O opened the pull request:
Fixes #174.

See: https://developer.android.com/training/system-ui/navigation
> When the user reopens the activity, onCreate() won't get called, so the system bars will remain visible. If you want system UI changes to persist as the user navigates in and out of your activity, set UI flags in onResume() or onWindowFocusChanged().

Maxr1998 reviewed (commented):
Good catch, thanks for the fix. Could you move the `onResume` function below `onStart` please? That order makes more sense imo considering the activity lifecycle.

ThreeFive-O commented:
@Maxr1998 Thank you for the feedback. Fixed the feedback and the code comment which was outdated.

One thing I noticed is that onResume doesn't fully fix the case when returning (maximizing) from Picture in Picture (PiP) mode. In most cases the navigation bar is still visible, but can be hidden by tapping the player window. In rare cases the navigation bar hides when returning from PiP.
Returning from PiP calls onResume, but there must be some other problem involved.
I'm on Android Pie on a Samsung phone.

Any idea?

Maxr1998 commented:
I've observed that as well when first implementing PiP, but never found a solution/fix for the issue. I'm thinking it's a bug in the Android framework or ExoPlayer library, but I can't really tell.

YouTube opens the video in portrait mode when returning from PiP, so the bug doesn't apply there. Maybe we should do the same to cover it, it also makes sense since you mostly watch through PiP while in portrait.

ThreeFive-O commented:
@Maxr1998 It took me a while to figure out the fullscreen logic correctly, but I guess with the latest commit I addressed all issues.

The behavior of the latest commit is as follows:

- When starting the black we request landscape orientation. The configuration change handler will take care about fullscreen mode.
- When playback is stopped and app is paused (because of an app switch) the `onResume` handler will restore the latest viewing mode (fullscreen when it was fullscreen; normal mode when it was normal mode)
- When switching to PiP mode the configuration changes for enabling/disabling fullscreen mode are not handled. This fixes the rogue orientation change in the configuration change handle.

Please give it a try.

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/player/PlayerFragment.kt:
```diff
@@ -180,6 +177,18 @@ class PlayerFragment : Fragment() {
+
+        // When returning from another app, the last active mode has to be set again
+        with(requireActivity()) {
+            if(isLandscape(resources.configuration))
+                enableFullscreen()
+            else
+                disableFullscreen()
```
```suggestion
            if (isLandscape()) enableFullscreen() else disableFullscreen()
```

Maxr1998 commented on app/src/main/java/org/jellyfin/mobile/player/PlayerFragment.kt:
```diff
@@ -204,6 +213,9 @@ class PlayerFragment : Fragment() {
      */
     private fun updateFullscreenStateAndSwitcher(configuration: Configuration) {
+        // Do not handle any orientation changes while being in Picture-in-Picture mode
+        if(requireActivity().isInPictureInPictureMode) return
```
```suggestion
        if (requireActivity().isInPictureInPictureMode)
            return
```
You could also combine all the calls to `requireActivity()` in this method to use a `with()` block.

ThreeFive-O replied:
Wrapped and done!

Maxr1998 reviewed (changes requested):
Thanks for your other set of changes, looks good to me! Just a small formatting nitpicks to fix, then it's ready to merge!

nielsvanvelzen commented on app/src/main/java/org/jellyfin/mobile/player/PlayerFragment.kt:
```diff
@@ -204,12 +210,18 @@ class PlayerFragment : Fragment() {
+                return
+
+            if (isLandscape(configuration)) enableFullscreen() else disableFullscreen()
+            val fullscreenDrawable = when {
+                !isFullscreen() -> R.drawable.ic_fullscreen_enter_white_32dp
+                else -> R.drawable.ic_fullscreen_exit_white_32dp
+            }
```
```suggestion
            val fullscreenDrawable = if (isFullscreen())
                R.drawable.ic_fullscreen_exit_white_32dp
            else
                R.drawable.ic_fullscreen_enter_white_32dp
```

Maxr1998 replied:
I'd personally keep it as is, even though the when isn't necessary here, it looks cleaner and makes it easier to see the two used drawable resources.

nielsvanvelzen replied:
In that case I'd invert the condition, instead of checking for `!isFullscreen()` use `isFullscreen()`. More readable that way.

ThreeFive-O commented:
Thank you @Maxr1998 and @nielsvanvelzen for your code reviews and improvements.

Maxr1998 commented:
Added another small fix (the `isPictureInPictureMode` call could have crashed below N) and rebased/squashen all commits onto the lastest master.
